src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs

```diff
@@ -16,6 +16,7 @@ class JsonMessageSerializer : IMessageSerializer
         Func<Stream, JsonReader> readerCreator;
         Func<Stream, JsonWriter> writerCreator;
         NewtonSerializer jsonSerializer;
+        static readonly Encoding utf8NoBOM = new UTF8Encoding(false);
```

Comment:
According to [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.-ctor?view=netframework-4.7.2#System_Text_UTF8Encoding__ctor_System_Boolean_):
> For security reasons, you should enable error detection by calling a constructor that includes a `throwOnInvalidBytes` parameter and setting its value to `true`.

Reply:
That would suggest we'd want the overload that takes two bools, but if you look at the `Encoding.UTF8` [docs](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.utf8?view=netframework-4.7.2#remarks), it turns invalid bytes into `?` instead.

It seems that the versions that you can create don't have that option. They either ignore invalid bytes or throw.

Neither option seems to be a good fit here, so I'm not sure what we can do here.

Reply:
If you didn't supply an encoder, this is what it would use:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/io/streamwriter.cs#L112

Which then defaults to false:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/text/utf8encoding.cs#L77

I guess to preserve the existing behavior, it doesn't throw and puts in ?'s. In the interest of preserving behavior, I'd see that.

I can add a test for this scenario, though.

Reply:
@jbogard It's not clear to me what you're suggesting. At least from the docs, it appears that there are three different behaviors regarding how invalid bytes are handled:
- `Encoding.UTF8`: replace with `?`
- `new UTF8Encoding (false, false)` : "ignore them"
- `new UTF8Encoding (false, true)`: "throw exception"

If that is the case, then this seems like a problem since it would cause an observable difference in how we handle them after this PR, and that gets into wire compat concerns.

Perhaps the `(false, false)` case is actually the same as `Encoding.UTF8` despite the docs not making that clear?

Reply:
If you *don't* supply an encoder, it's `(false, true)`. If you use `Encoding.UTF8`, it's `(true, false)`:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/text/encoding.cs#L1549

To preserve the existing behavior on throwing, we'd want `false` for the second parameter, for writing. I left the existing reading encoder as-is.

So I think we'd want `(false, false)`, changing the BOM for writing but leaving the 2nd argument as-is.

Reply:
I'm going to add the second argument to make it explicit.

Reply:
For posterity, `StreamReader` defaults to `Encoding.UTF8`:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/io/streamreader.cs#L138

Which is `(true, false)`.

Reply:
Am I supposed to mark the conversation as resolved?